---
title: "Basic usage of snpEnrichR package"
author:
- name: Kari Nousiainen
  affiliation: 
  - Aalto Univerisity School of Science
  email: Kari.Nousiainen@Aalto.FI
- name: Kartiek Kanduri
  affiliation: 
    - Aalto Univerisity School of Science
    - University of Turku
  email: Kartiek.Kanduri@Aalto.FI 
date: "`r Sys.Date()`"
package: snpEnrichR
output:   BiocStyle::html_document
vignette: > 
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

<!--
%% \VignetteEngine{knitr::knitr}
-->

```{r style, echo = FALSE, results = 'asis'}
BiocStyle::markdown()
```


# Abstract

snpEnrichR implements methods to analyze enrichment of SNPs in genomic regions.

# Supported functionalities
- Download SNPs from EBI's GWAS
- Decorrelate  SNP sets
- Query randomized SNP sets matching the SNP set from snpSNAP
- Find SNPs correlated to input SNPs within a given distance
- Perform enrichment analysis, including overlap computation, empirical statistical hypothesis testing with Benjamini-Hochberg multiple hypothesis correction 

# Citation 

The citation instructions come here when published.

# Introduction

Associations between SNPs and genomic regions can provide understanding about the  molecular mechanics underlying any given phenotype. snpEnrichR is a tool for estimating Enrichment of such association. In order to estimate a null distribution for empirical  statistical hypothesis test the set of genomic regions is considered as the reference set, and the set of trait specific SNPs is then randomized multiple times. 




Genome-wide  association  studies  have  successfully  linked  SNPs  to
various traits. 

Resolution of such studies    

imited resolution of genotyping methods
used in GWAS studies and because SNPs occur in linkage disequilibrium
(LD), it is possible that the link between the phenotype and the detect


However, due to a limited resolution of genotyping methods
used in GWAS studies and because SNPs occur in linkage disequilibrium
(LD), it is possible that the link between the phenotype and the detecte

lead SNP is indirect through another SNP, which is in LD with the lead
SNP. In these cases,  taking proxy SNPs into account may enhance the
sensitivity of the enrichment analysi


Vignettes are long form documentation commonly included in packages. Because they are part of the distribution of the package, they need to be as compact as possible. The `html_vignette` output type provides a custom style sheet (and tweaks some options) to ensure that the resulting html is as small as possible. The `html_vignette` format:

- Never uses retina figures
- Has a smaller default figure size
- Uses a custom CSS stylesheet instead of the default Twitter Bootstrap style

## Vignette Info

Note the various macros within the `vignette` section of the metadata block above. These are required in order to instruct R how to build the vignette. Note that you should change the `title` field and the `\VignetteIndexEntry` to match the title of your vignette.

## Styles

The `html_vignette` template includes a basic CSS theme. To override this theme you can specify your own CSS in the document metadata as follows:

    output: 
      rmarkdown::html_vignette:
        css: mystyles.css

## Figures

The figure sizes have been customised so that you can easily put two images side-by-side. 

```{r, fig.show='hold'}
plot(1:10)
plot(10:1)
```

You can enable figure captions by `fig_caption: yes` in YAML:

    output:
      rmarkdown::html_vignette:
        fig_caption: yes

Then you can use the chunk option `fig.cap = "Your figure caption."` in **knitr**.

## More Examples

You can write math expressions, e.g. $Y = X\beta + \epsilon$, footnotes^[A footnote here.], and tables, e.g. using `knitr::kable()`.

```{r, echo=FALSE, results='asis'}
knitr::kable(head(mtcars, 10))
```

Also a quote using `>`:

> "He who gives up [code] safety for [code] speed deserves neither."
([via](https://twitter.com/hadleywickham/status/504368538874703872))
