---
title: "Basic usage of snpEnrichR package"
author:
- name: Kari Nousiainen
  affiliation: 
  - Aalto Univerisity School of Science
  email: Kari.Nousiainen@Aalto.FI
- name: Kartiek Kanduri
  affiliation: 
    - Aalto Univerisity School of Science
    - University of Turku
  email: Kartiek.Kanduri@Aalto.FI 
date: "`r Sys.Date()`"
package: snpEnrichR
output:   BiocStyle::html_document
vignette: > 
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

<!--
%% \VignetteEngine{knitr::knitr}
-->

```{r style, echo = FALSE, results = 'asis'}
BiocStyle::markdown()
```


# Abstract

snpEnrichR implements methods to analyze enrichment of SNPs in genomic regions. This document shows how to do co-localization analysis using snpEnrichR functions. 

# Supported functionalities
* Download SNPs from EBI's GWAS
* Decorrelate  SNP sets
* Query randomized SNP sets matching the SNP set from snpSNAP
* Find SNPs correlated to input SNPs within a given distance
* Perform enrichment analysis, including overlap computation, empirical statistical hypothesis testing with Benjamini-Hochberg multiple hypothesis correction 

# Citation 

The citation instructions come here when published.

# Introduction

snpEnrichR is a tool for estimating Enrichment of association between (trait specific) SNPs and genomic regions by performing an empirical statististical test. In order to estimate a null distribution for a statistical hypothesis test the set of genomic regions is considered as the reference set, and the set of trait specific SNPs is randomized multiple times such that the major characteristics of SNPs are conserved. As SNPs occur in linkage disequilibrium (LD) and genotyping methods used in GWAS studies have limited resolution, the link between SNPs and genomic regions may be indirect; observed through other SNPs in the LD. snpEnrichR supports possibility to expand  the set of SNPs with their LD buddies.

# `getSNPs`: Download trait specific SNPs from NHGRI-EBI GWAS Catalog 
snpEnrichR provides `getSNPs` for NHGRI-EBI GWAS Catalog query.
```R
# query term
qt <- 'psoriasis'
GWAS_SNPS <- getSNPs(qt)
str(GWAS_SNPS,vec.len=1)
```

```R
'data.frame':	463 obs. of  34 variables:
 $ DATE ADDED TO CATALOG     : chr  "2010-12-01" ...
 $ PUBMEDID                  : chr  "20953189" ...
 $ FIRST AUTHOR              : chr  "Stuart PE" ...
 $ DATE                      : chr  "2010-10-17" ...
 $ JOURNAL                   : chr  "Nat Genet" ...
 $ LINK                      : chr  "www.ncbi.nlm.nih.gov/pubmed/20953189" ...
 $ STUDY                     : chr  "Genome-wide association analysis identifies three psoriasis susceptibility loci." ...
 $ DISEASE/TRAIT             : chr  "Psoriasis" ...
 $ INITIAL SAMPLE SIZE       : chr  "Up to 1,831 cases, up to 2,546 controls" ...
 $ REPLICATION SAMPLE SIZE   : chr  "Up to 4,064 cases, up to 4,685 controls" ...
 $ REGION                    : chr  "17q11.2" ...
 $ CHR_ID                    : chr  "17" ...
 $ CHR_POS                   : chr  "27779649" ...
 $ REPORTED GENE(S)          : chr  "NOS2" ...
 $ MAPPED_GENE               : chr  "NOS2" ...
 $ UPSTREAM_GENE_ID          : chr  "" ...
 $ DOWNSTREAM_GENE_ID        : chr  "" ...
 $ SNP_GENE_IDS              : chr  "4843" ...
 $ UPSTREAM_GENE_DISTANCE    : chr  "" ...
 $ DOWNSTREAM_GENE_DISTANCE  : chr  "" ...
 $ STRONGEST SNP-RISK ALLELE : chr  "rs4795067-G" ...
 $ SNPS                      : chr  "rs4795067" ...
 $ MERGED                    : chr  "0" ...
 $ SNP_ID_CURRENT            : chr  "4795067" ...
 $ CONTEXT                   : chr  "intron_variant" ...
 $ INTERGENIC                : chr  "0" ...
 $ RISK ALLELE FREQUENCY     : chr  "0.35" ...
 $ P-VALUE                   : chr  "4E-11" ...
 $ PVALUE_MLOG               : chr  "10.397940008672037" ...
 $ P-VALUE (TEXT)            : chr  "" ...
 $ OR or BETA                : chr  "1.19" ...
 $ 95% CI (TEXT)             : chr  "[NR]" ...
 $ PLATFORM [SNPS PASSING QC]: chr  "Illumina, Perlegen [up to 7456344] (imputed)" ...
 $ CNV                       : chr  "N" ...
```
The data frame contains all SNPs available. Hence, it should be parsed to match required criteria e.g. disease/trait, population, p-value etc. For example, considering all SNPs where Disease/treat is "Psoriasis" related to studies where cases and control are from European populations and keeping duplicate SNPs in the list would produce the following table.

```R
'data.frame':	55 obs. of  5 variables:
 $ CHR_ID       : chr  "13" "6" "5" "5" ...
 $ CHR_POS      : chr  "39776775" "31285148" "159290781" "151098757" ...
 $ DISEASE/TRAIT: chr  "Psoriasis" "Psoriasis" "Psoriasis" "Psoriasis" ...
 $ P-VALUE      : chr  "2E-6" "1E-100" "2E-28" "1E-20" ...
 $ SNPS         : chr  "rs7993214" "rs12191877" "rs2082412" "rs17728338" ...
```
The SNPs can be extracted from the snplist.  

```R
snplist <- unlist(apply(selected.snps[,1:2],1,function(x) base::paste0(x,collapse =':')),use.names = F)
names(snplist) <- NULL
```
The list must not contain several SNPs in a LD block as it would bias the randomly generated SNP sets. We use clumpSNPs to remove extra SNPs from the list.


# `clumpSNPs`: Decorrelate  SNP sets 
SNPs must be decorrelated as correlated SNPs cause biases when drawing random SNP sets.
```R
snplist.nc <- clumpSNPs('Data/ThousandGenomes/1kp3_snpSNAP',
                          snplist,
                          '',
                          0.8,1000,
                          ChainFilePath='Data/hg38ToHg19.over.chain')
```

# `submitSNPsnap`: Query randomized SNP sets matching the SNP set from snpSNAP
submitSNPsnap queries snpSNAP server to produce random sets of SNPs matching input SNPs with given criteria.  
```R
snpSNAPUrl = submitSNPsnap(snplist.nc,
                          distance_type ="kb",
                          distance=1000,
                          clump_r2="0.8",
                          clump_kb="1000",
                          max_ld_buddy_count_deviation=20,
                          ld_buddy_cutoff = "0.8",
                          N_sample_sets=1000,
                          email_address = "my@address",
                          job_name="PS")
```
When available, the snpSNAP results can be downloaded.
```R
  download.file(snpSNAPUrl,
                "SNPSnap_Psoriasis_download",
                method="wget",
                quiet = FALSE, 
                mode = "w")
  unzip(zipfile="SNPSnap_Psoriasis_download",exdir="data/PS",junkpaths=T)
```

In addition to randomized SNP sets, SNPsnap output files contain information needed for quality control. It is important to confirm everything is order before continuing.
```R
dir("data/PS/")
```
```
[1] "input_snps_clumped.txt"              "input_snps_excluded.txt"            
[3] "input_snps_identifer_mapping.txt"    "input_snps_insufficient_matches.txt"
[5] "matched_snps.txt"                    "snpsnap_summary.txt"    
```


```R
matchedSnps <- unlist(read.table('data/PS/matched_snps.txt',header=T,sep="\t",stringsAsFactors = F),use.names = F)
```

# `findProxies`: Find LD buddies for SNPs
 `findProxies` computes proxy SNPs for given SNP list

```R
findProxies(path2PlinkPrefix="Data/ThousandGenomes/1kp3_snpSNAP",
            snplist=matchedSnps,
            path2Proxies="data/Psoriasis/proxies/PS")

```
# `analyzeEnrichment`: Perform the actual analysis

```R
      analyzeEnrichment(regionPath="data/macs_peaks",
                    regionHeader=c('chr','start','end'),
                    SNPsnapPath="data/",
                    numberOfRandomSNPsets=1000,
                    proxyPathPrefix="data/Psoriasis/proxies/",
                    traitShort='PS',
                    genomicRegionsName='macs peaks',
                    cores=4,
                    resDir="data",
                    traitsLong='Psoriasis',
                    ChainFilePath="Data/hg38ToHg19.over.chain")
```

AnalyzeEnrichment produces files disease_enrichment.txt, Overlaps_PS.txt and PS_background_distributions.png into the PS_background_distributions.png. They contain the p-value, the numbers of the overlaps between the SNP sets and the genomic regions and the histogram of the overlaps, respectively.   
<!--
Vignettes are long form documentation commonly included in packages. Because they are part of the distribution of the package, they need to be as compact as possible. The `html_vignette` output type provides a custom style sheet (and tweaks some options) to ensure that the resulting html is as small as possible. The `html_vignette` format:

- Never uses retina figures
- Has a smaller default figure size
- Uses a custom CSS stylesheet instead of the default Twitter Bootstrap style

## Vignette Info

Note the various macros within the `vignette` section of the metadata block above. These are required in order to instruct R how to build the vignette. Note that you should change the `title` field and the `\VignetteIndexEntry` to match the title of your vignette.

## Styles

The `html_vignette` template includes a basic CSS theme. To override this theme you can specify your own CSS in the document metadata as follows:

    output: 
      rmarkdown::html_vignette:
        css: mystyles.css

## Figures

The figure sizes have been customised so that you can easily put two images side-by-side. 

```{r, fig.show='hold'}
plot(1:10)
plot(10:1)
```

You can enable figure captions by `fig_caption: yes` in YAML:

    output:
      rmarkdown::html_vignette:
        fig_caption: yes

Then you can use the chunk option `fig.cap = "Your figure caption."` in **knitr**.

## More Examples

You can write math expressions, e.g. $Y = X\beta + \epsilon$, footnotes^[A footnote here.], and tables, e.g. using `knitr::kable()`.

```{r, echo=FALSE, results='asis'}
knitr::kable(head(mtcars, 10))
```

Also a quote using `>`:

> "He who gives up [code] safety for [code] speed deserves neither."
([via](https://twitter.com/hadleywickham/status/504368538874703872))

-->
